test/db/analysis/vars

```diff
@@ -590,19 +592,21 @@ EOF
 EXPECT=<<EOF
             ;-- varfunc:
 / dbg.varfunc ();
-|           ; var int lightbulb @ stack - 0x10
-|           ; var int sun @ stack - 0xc
-|           ; var int last @ stack - 0x8
-|           ; var int chance @ stack - 0x4
+|           ; var int lightbulb @ stack - 0x18
+|           ; var int sun @ stack - 0x14
+|           ; var int last @ stack - 0x10
+|           ; var int chance @ stack - 0xc
+|           ; var int64_t var_8h @ stack - 0x8
+|           ; var int64_t var_4h @ stack - 0x4
 |           0x00001137      sub   rsp, 0x10
 |           0x0000113b      mov   dword [rsp], 1
-|           0x00001142      mov   dword [sun], 2
-|           0x0000114a      mov   dword [last], 3
-|           0x00001152      mov   dword [chance], 4
+|           0x00001142      mov   dword [chance], 2
+|           0x0000114a      mov   dword [var_8h], 3
+|           0x00001152      mov   dword [var_4h], 4
```

Comment:
this is still wrong according to IDA:
```
public varfunc
varfunc proc near

lightbulb= dword ptr -10h
sun= dword ptr -0Ch
last= dword ptr -8
chance= dword ptr -4

; __unwind {
sub     rsp, 10h
mov     [rsp+10h+lightbulb], 1
mov     [rsp+10h+sun], 2
mov     [rsp+10h+last], 3
mov     [rsp+10h+chance], 4
mov     eax, 0
call    leaffunc
mov     eax, [rsp+10h+chance]
mov     [rsp+10h+sun], eax
mov     eax, [rsp+10h+lightbulb]
add     rsp, 10h
retn
; } // starts at 1137
varfunc endp
```

maybe we should at least mark this as BROKEN

Reply:
disassemble from llvm-objdump:
```
000000000000113c <varfunc>:
    113c: 55                           	pushq	%rbp
    113d: 48 89 e5                     	movq	%rsp, %rbp
    1140: 48 83 ec 10                  	subq	$0x10, %rsp
    1144: c7 45 f0 01 00 00 00         	movl	$0x1, -0x10(%rbp)
    114b: c7 45 f4 02 00 00 00         	movl	$0x2, -0xc(%rbp)
    1152: c7 45 f8 03 00 00 00         	movl	$0x3, -0x8(%rbp)
    1159: c7 45 fc 04 00 00 00         	movl	$0x4, -0x4(%rbp)
    1160: b8 00 00 00 00               	movl	$0x0, %eax
    1165: e8 cb ff ff ff               	callq	0x1135 <leaffunc>
    116a: 8b 45 fc                     	movl	-0x4(%rbp), %eax
    116d: 89 45 f4                     	movl	%eax, -0xc(%rbp)
    1170: 8b 45 f0                     	movl	-0x10(%rbp), %eax
    1173: c9                           	leave
    1174: c3                           	retq
```

Reply:
In the current version of Rizin, it appears that the information from `.eh_frame` and `.debug_frame` is not being used to correct the position of variables.

Reply:
`0x113b: CFA=RSP+24: RIP=[CFA-8]`: From this information, I can infer that the value of CFA (Canonical Frame Address) has an offset of +24 (0x18).

However, in the disassembly line `1137: 48 83 ec 10 subq $16, %rsp`, we only subtracted 16 from RSP. I still don't know how the value of 24 is calculated.

from chatgpt:

> The discrepancy between the offset value in the CFA calculation and the subtraction in the disassembly can be explained by the fact that the CFA value is typically established based on the function prologue and the specific compiler optimizations or stack frame layout being used.
> 
> In the given disassembly line, `subq $16, %rsp`, we can see that 16 bytes are subtracted from the RSP register. This instruction typically reserves space on the stack for local variables or function parameters.
> 
> The additional offset of 8 bytes in the CFA calculation (`RIP=[CFA-8]`) may be due to other factors such as the storage of the return address or the size of certain data structures in the stack frame.
> 
> To accurately determine how the value of 24 is calculated in this context, a deeper analysis of the surrounding code and the specific compiler settings would be required.

Reply:
I have an idea. Perhaps we can introduce a new `RZ_ANALYSIS_VAR_STORAGE_CFA`, to represent the storage of variables.

During the disassembly of a function (`pdf`), we can output the following content:

```
/ dbg.varfunc ();
|           ; var int lightbulb @ CFA - 0x18
|           ; var int sun @ CFA - 0x14
|           ; var int last @ CFA - 0x10
|           ; var int chance @ CFA - 0xc
            ; CFA=RSP+8: RIP=[CFA-8]
|           0x00001137      sub   rsp, 0x10
            ; CFA=RSP+0x18: RIP=[CFA-8]
|           0x0000113b      mov   dword [lightbulb], 1
|           0x00001142      mov   dword [sun], 2
|           0x0000114a      mov   dword [last], 3
|           0x00001152      mov   dword [chance], 4
...
```

Reply:
@imbillow maybe these could be helpful?

- https://www.polarsignals.com/blog/posts/2022/11/29/dwarf-based-stack-walking-using-ebpf/
- https://dl.acm.org/doi/pdf/10.1145/3360572
- https://stackoverflow.com/questions/7534420/gas-explanation-of-cfi-def-cfa-offset

If it doesn't help, gathering more data from real-world binaries is a good idea; then, we can deal with this later once we better understand it. @thestr4ng3r, what do you think about this, since you are the one who migrated variables to the frame pointer?